### PR TITLE
Fix refilling of vehicle tanks

### LINKED ISSUE #17304: Refilling cars from containers broken
1 - Start new world
2 - Spawn a gas/diesel vehicle, rubber hose, gallon jug
3 - Siphon out some fuel
4 - Attempt to refill it

### PR DESCRIPTION
**Pull request by mugling**
Fixes #17304

### CODE REVIEW

**Comment by mugling** [comment]
Done. I did have an enhanced version that supported reloading a variable quantity of fuel from carried or adjacent containers via a menu but it would depended upon serialization of `item_location` (#16791) within `player_activity` which isn't yet ready so this fix should suffice for now.

File: src/veh_interact.cpp [hunk 1]
```diff
@@ -316,30 +315,19 @@ task_reason veh_interact::cant_do (char mode)
...
-            }
-        }
 
-        // if refillable parts exist but can't be refilled for some reason.
-        if (has_ptank) {
-            return CANT_REFILL;
-        }
+    case 'f': {
+        // first find all possible ammo
+        auto opts = g->u.items_with( []( const item &e ) { return e.is_ammo(); } );
```
Comment (BevapDin) [inline src/veh_interact.cpp]: Why only ammo? One should be able to create and fill a tank of bleach or salt water or any other liquid substance.

File: src/vehicle.cpp [hunk 2]
```diff
@@ -6290,6 +6341,26 @@ bool vehicle_part::is_light() const
            vp.has_flag( VPFLAG_ATOMIC_LIGHT );
 }
 
+bool vehicle_part::is_tank() const
+{
+    if( !info().has_flag( VPFLAG_FUEL_TANK ) ) {
+        return false;
+    }
+
+    auto *fuel = item::find_type( default_ammo( info().fuel_type ) );
```
Comment (BevapDin) [inline src/vehicle.cpp]: `info().fuel_type` is already an item type id, no need to call `default_ammo`.

File: src/vehicle.cpp [hunk 3]
```diff
@@ -6154,6 +6150,28 @@ std::string vehicle_part::name() const {
     return res;
 }
 
+ammotype vehicle_part::ammo_type() const
+{
+    // @todo generic fuel tanks are not yet supported
+    if( is_tank() ) {
+        return info().fuel_type;
```
Comment (BevapDin) [inline src/vehicle.cpp]: Won't work for example for napalm tanks. It would return the item type id of napalm, but that is not an ammo type id.

File: src/veh_interact.cpp (hunk 1 above)
Comment (mugling) [inline src/veh_interact.cpp]: The groundwork for that is set via 236cea2 and fc71a44.
I did have a more complex version that supported arbitrary reloading but I've scaled it back to just sufficient for fixing this bug for now.

File: src/vehicle.cpp (hunk 3 above)
Comment (mugling) [inline src/vehicle.cpp]: Good point

File: src/veh_interact.cpp (hunk 1 above)
Comment (BevapDin) [inline src/veh_interact.cpp]: I can (right now, I tested it) copy the definition of a water tank vehicle part in JSON and replace `"fuel_type": "water"` with `"fuel_type": "bleach"` and have a bleach tank. I can fill it with bleach just fine (by unloading a container).

Checking for `fuel_type` being ammo would (at best) lead to inconstant behavior (unloading allows liquids, refilling via vehicle menu allows ammo only).

Comment (mugling) [inline src/veh_interact.cpp]: `bleach` doesn't have an `ammo_type` so this won't work after 3047635.

For now fuel tanks have to contain ammo; this is sufficient to both fix the bug and is not a regression from `master`.

Generic vehicle tanks requires more support and testing and can be considered as part of a separate PR.

Comment (BevapDin) [inline src/veh_interact.cpp]: > bleach  doesn't have an  ammo_type  so this won't work after  3047635 .

Exactly. A regression: something was possible before and is not possible afterwards.

> For now fuel tanks have to contain ammo; this is sufficient to both fix the bug and is not a regression from  master .

It would probably not harm to change the check to something like `e.is_ammo() || e.made_of(LIQUID)` and thereby keep the existing behavior.

> Generic vehicle tanks requires more support and testing and can be considered as part of a separate PR.

This is not about generic tanks. This is about the existing tank system.

Comment (Coolthulhu) [inline src/veh_interact.cpp]: > bleach doesn't have an ammo_type so this won't work after 3047635.

Add it to consistency checks. Then `--check-mods`.

Comment (mugling) [inline src/veh_interact.cpp]: We don't actually have bleach tanks however or indeed tanks that hold anything other than ammo. This and any future implementation is greatly simplified by following this rule as opposed to supporting arbitrary functionality that we currently don't use. YAGNI.

If we we did want bleach tanks we should support them the same way we do water tanks, by providing an `ammo` field in the JSON definition.

Comment (mugling) [inline src/veh_interact.cpp]: Confirmed via consistency check that all `VP_FUEL_TANK` parts only hold `ammo`

Comment (BevapDin) [inline src/veh_interact.cpp]: > We don't actually have bleach tanks however or indeed tanks that hold anything other than ammo. This and any future implementation is greatly simplified by following this rule as opposed to supporting arbitrary functionality that we currently don't use. YAGNI.

YAGNI is my argument as well. What part of the ammo data is required by the tank vehicle part? Container items don't require their content to be ammo. Tank vehicle parts should behave similar.

> If we we did want bleach tanks we should support them the same way we do water tanks, by providing an  ammo  field in the JSON definition.

Now you're advocating a hack. Again, why does a liquid need ammo data just to store it in a vehicle tank?

Comment (mugling) [inline src/veh_interact.cpp]: > What part of the ammo data is required by the tank vehicle part?

The contents must be counted by charges and therefore either `AMMO` or `COMESTIBLE`.

> YAGNI is my argument as well.

Your suggestion enables support for something that we don't make any use of either in mainline  content or indeed any mods.

> why does a liquid need ammo data just to store it in a vehicle tank?

Yes, with time we probably need to separate out `AMMO` into it's different facets but this affects pretty much the entire codebase. That wouldn't be appropriate for a bugfix.

Comment (mugling) [inline src/veh_interact.cpp]: Additionally I'm not sure what corner cases exist for non-ammo contents of fuel tanks. It possibly works but it probably doesn't and in the absence of any actual uses it's safer to document it as a precondition via the consistency check and then later relax that if necessary rather than the opposite.

Comment (BevapDin) [inline src/veh_interact.cpp]: > The contents must be counted by charges and therefore either  AMMO  or  COMESTIBLE .

Then why are you testing for it being ammo? If you need the count-by-charges property, test for it. There are suitable static and non-static functions in the item class.

> Additionally I'm not sure what corner cases exist for non-ammo contents of fuel tanks. It possibly works but it probably doesn't and in the absence of any actual uses it's safer to document it as a precondition via the consistency check and then later relax that if necessary rather than the opposite.

You're picking one property (being ammo) of the currently used fuel types and make it mandatory. But what about the other properties? Are you sure there are no corner cases that require some other property as well?

If you're worried about some corner cases, restrict the fuel types to only those that are known to be supported.

> Your suggestion enables support for something that we don't make any use of either in mainline content or indeed any mods.

I *keeps* support for that. It does not enable it, it's already there.

Comment (mugling) [inline src/veh_interact.cpp]: > It does not enable it, it's already there.

But it is entirely unused. If it it works, completely and without bugs, it's purely by chance as opposed to conscious design. Given sufficient time such support will only bitrot. Sure we could write test cases to prevent this, but how is that not YAGNI - we don't use it now nor is there a clear need to support it.

This versus a precondition that allows the other code to be simpler. You can always extend from that if you do decide you need it.

> Container items don't require their content to be ammo.

We probably should work on container mechanics to allow restriction on their contents. At that time they can become a direct replacement for vehicle tanks.

Comment (BevapDin) [inline src/veh_interact.cpp]: > Turrets expect contents of fuel tanks to have an  ammo_type

How so? Code for example in `vehicle::turret_reload` explicitly allows fuel (as reported by `vehicle::fuels_left`) to be non-ammo. It checks `itype::ammo` being non-null and ignores fuel types that aren't ammo:
```C++
    for( const auto &e : fuels_left() ) {
        const itype *fuel = item::find_type( e.first );
        if( !fuel->ammo || e.second < gun.ammo_required() ) {
            continue;
        }
        ...
```

It seems non-ammo fuel types are supported just fine.

Comment (mugling) [inline src/veh_interact.cpp]: Good point , I did forget I wrote that although it was with the intention of being certain to never get a nullptr deference. We'd have to check all of the remaining turret code to be sure.

How do you propose to stop such *support* bitrotting? Emphasis mine but in the lack of any actual in-the-wild use examples I think that's a fair assessment.

If we allow non-ammo what should `ammo_current()` and `ammo_type()` return for non-ammo fuels? 

I want to avoid the special case where `ammo_current() == "null"` but `ammo_remaining() > 0` if at all possible. The exists in the `item` code due to legacy items (spray can, lighter etc) but has been the cause of a lot of bugs and adds a lot of boilerplate to the code.

Comment (mugling) [inline src/veh_interact.cpp]: > I want to avoid the special case

Or even worse `ammo_current() != "null"` but `ammo_type() == "NULL"`. That doesn't occur anywhere else and the `vehicle::ammo_*` functions having different semantics so as to support something we make no use of is bad design/YAGNI.

Comment (BevapDin) [inline src/veh_interact.cpp]: > How do you propose to stop such support bitrotting? Emphasis mine but in the lack of any actual in-the-wild use examples I think that's a fair assessment.

We don't need to. It works fine. If you're worried, add a new vehicle tank part that uses for example salt water as fuel type and the feature is not unused anymore.

I have not seen any good reason why the fuel type *needs* the restriction. Especially as it encourages other hacks (the mentioned addition of ammo data to allow storing bleach in vehicle tanks).

> If we allow non-ammo what should  ammo_current()  and  ammo_type()  return for non-ammo fuels?

> I want to avoid the special case where  ammo_current() == "null"  but  ammo_remaining() > 0  if at all possible. The exists in the  item  code due to legacy items (spray can, lighter etc) but has been the cause of a lot of bugs and adds a lot of boilerplate to the code.

Why not add a `content_count(ammotype)` function (different name maybe), which returns the charges of the content (if the content is ammo and has a matching ammo type) or 0 otherwise.

And with `ammotype` being a separate type, you can overload the function to take a `itype_id` to check how much of a given item it contains.

And maybe a third overload without parameters to get the content charges of any content type.

--

Regarding items like spray can or lighter: we probably have to decide - do they contain ammo or do they only contain "charges" (an arbitrary number of usages)?

If ammo, they should get a proper ammo type (and flags to prevent unloading/reloading as needed). Cash cards for example use this.

If "charges", add a `get_charges` function (or `get_remaining_usages`) that returns exactly that. Similar to `get_remaining_chapters` for books.

Comment (mugling) [inline src/veh_interact.cpp]: @BevapDin ok. I've gone with your suggestion for now (5e7a41d) and in reviewing the code have a clearer plan for generic tanks but I'm going to do this in a separate PR. Presuming from the above that you've had a good chance to check this PR so are you happy to merge test it so I can carry on with part 2?

**Comment by Coolthulhu** [comment]
Minor conflict with #17351

**Comment by Coolthulhu** [comment]
I'm getting a debug spam of "can_reload was supplied with non-ammo".

**Comment by mugling** [comment]
Fixed both merge conflicts and above issue

File: src/veh_interact.cpp [hunk 4]
```diff
@@ -957,6 +944,16 @@ void veh_interact::do_refill()
         break; // no reason, all is well
     }
 
+    std::vector<vehicle_part *> ptanks;
+    auto opts = g->u.items_with( []( const item &e ) { return e.is_ammo(); } );
```
Comment (BevapDin) [inline src/veh_interact.cpp]: This calls `is_ammo`, the other loop in `cant_do` calls `count_by_charges`, should be the same in both places.

In fact, why not move this loop into a function that returns the `ptanks` vector?

**Comment by Coolthulhu** [comment]
Minireactor math is off - installing, loading, removing and unloading will double the number of plutonium cells.

**Comment by Coolthulhu** [comment]
Improvement over master, but that minireactor exploit needs fixing.
